### PR TITLE
Render Sphinx `/docs` in docusaurus

### PR DESCRIPTION
**Pull request by jaimergp**
PR Checklist:

- [ ] note any issues closed by this PR with [closing keywords](https://help.github.com/articles/closing-issues-using-keywords)
- [x] put any other relevant information below

---

- Part of #1971.
- #2066 and #2068 should probably go in before this one.
- Based on https://github.com/Quansight-Labs/cf-infra-docs/pull/14. 
- How it works:
  - This uses a fork of `sphinx-markdown-builder` that adds better Docusaurus support (e.g. admonitions) to render the Sphinx docs to markdown documents.
  - The rendered Markdown has resolved all the references and Sphinx directives to regular Markdown Docusaurus can understand.
  - These MD documents are moved to `/docs` by a special script that also handles some minor postprocessing (adding `sidebar_position`, rewriting some URLs, etc).
  - From there, Docusaurus will take `/docs` as if they were written by hand and publish the site.
- The default URL structure is a bit different in docusaurus. Directory names like `/path/thing/` are preferred to file names like `/path/thing.html`. Our Sphinx site also had strange `00_intro.html` instead of `index.html`. To that effect, we have added redirects from `/path/to/thing.html -> /path/to/thing/` and `/path/00_intro.html -> /path/`.
- Since we have now all the site rendered under Docusaurus, we can add some plugins:
  - Algolia search for super awesome search results. These should become available once we merge. It should look like the search bar in conda.org. If we don't want Algolia, we can use `lunr` (local search, JS plugin, not as powerful).
  - GoatCounter for OSS-friendly stats, no cookies 

Notes for reviewers:

- We have PR previews now! Check how the new docs look like at https://deploy-preview-2077--conda-forge-previews.netlify.app/docs/.
- There might be minor formatting issues so hopefully we catch those during the review. Screenshots or links in the preview would be super helpful!
- Yes, the `sphinx-markdown-builder` fork situation is a bit fragile. We can try to upstream things but: (1) some things are Docusaurus specific and might need options; (2) this is supposed to be temporary to some degree, because ideally we would end up writing everything in native Docusaurus 🤞. If (2) doesn't happen, then yes, upstreaming is even more important.

### CODE REVIEW

File: .ci_scripts/sphinx_markdown_to_docusaurus.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+"""
+import sys
+from pathlib import Path
+
+from ruamel.yaml import YAML
+
+yaml = YAML(typ="safe")
+yaml.default_flow_style = False
+
+SIDEBAR_ORDERING = [
```
Comment (beckermr) [inline .ci_scripts/sphinx_markdown_to_docusaurus.py:12]: If we change the website's high-level structure, does that mean this list will need to change too? If yes, we should add a comment and maybe have some sort of CI check.

File: .ci_scripts/update_docs [hunk 2]
```diff
@@ -2,35 +2,27 @@
...
 
-python .ci_scripts/generate_cfep_index.py
-
-pushd sphinx/src
+python "$BASE_DIR/.ci_scripts/generate_cfep_index.py"
 
+pushd "$BASE_DIR/sphinx/src"
+make clean
 # -W --keep-going: list all warnings but fail build in case there are any
-# -n: check validity of all links
```
Comment (beckermr) [inline .ci_scripts/update_docs:17]: I guess we cannot run link check anymore? This has caught many bugs for us.

**Review by beckermr — COMMENTED** [review]
LGTM from me but I left a few comments.

File: .ci_scripts/update_docs (hunk 2 above)
Comment (jaimergp) [reply .ci_scripts/update_docs:17]: Docusaurus runs it later (and much faster), so no need to do it with Sphinx. Let me push a broken link on purpose to show it in action.

Comment (beckermr) [reply .ci_scripts/update_docs:17]: No need to test it. I didn't know! Thank you!

Comment (jaimergp) [reply .ci_scripts/update_docs:17]: Well, turns out it only check internal links, no external stuff. Adding a new step for now.

File: .ci_scripts/sphinx_markdown_to_docusaurus.py (hunk 1 above)
Comment (jaimergp) [reply .ci_scripts/sphinx_markdown_to_docusaurus.py:9]: Good catch! Added a check to raise if the document is not found in this list. Hopefully, if we start changing `sphinx/` stuff, is to bring it over to Docusaurus, so "this should not happen" but better be safe than sorry.

**Comment by beckermr** [comment]
Well I tried to fix the core and emeritus member lists @jaimergp but I think I am missing something important about how sphinx is working now or I got the relative path wrong.

File: sphinx/src/conf.py [hunk 3]
```diff
@@ -355,10 +358,10 @@ def get_formated_names(path_file):
     current_file = os.path.dirname(__file__)
     context = app.config.html_context
     context["core_members"] = get_formated_names(
-        os.path.join(current_file, "..", "..", "src", "core.csv")
+        os.path.join(current_file, "..", "..", "..", "src", "core.csv")
```
Comment (beckermr) [inline sphinx/src/conf.py:361]:
```suggestion
        os.path.join(current_file, "..", "..", "..", "..", "src", "core.csv")
```

File: sphinx/src/conf.py [hunk 4]
```diff
@@ -355,10 +358,10 @@ def get_formated_names(path_file):
...
     )
     context["emeritus_members"] = get_formated_names(
-        os.path.join(current_file, "..", "..", "src", "emeritus.csv")
+        os.path.join(current_file, "..", "..", "..", "src", "emeritus.csv")
```
Comment (beckermr) [inline sphinx/src/conf.py:364]:
```suggestion
        os.path.join(current_file, "..", "..", "..", "..", "src", "emeritus.csv")
```

**Comment by jaimergp** [comment]
Let me take a look, you mean the governance.rst substitution, right?

**Comment by jaimergp** [comment]
Ok, the early return branch made us skip that part because the rendered is not HTML anymore, but Markdown. I adjusted a couple things and also added the CFEP substitution there because it's cleaner (otherwise you might risk accidentally committing the rendered version, like it happened two days ago :D).

**Comment by jaimergp** [comment]
Fixed! See:

* https://deploy-preview-2077--conda-forge-previews.netlify.app/docs/orga/governance/#current-members-of-core
* https://deploy-preview-2077--conda-forge-previews.netlify.app/docs/orga/cfep-index/

**Comment by beckermr** [comment]
Awesome. This last thing was the only comment on slack. I'd say lgtm and let's merge this. Any other issues can be fixed later.

**Comment by jaimergp** [comment]
Yes, I can take a look on the search bar too. Hopefully it starts working once we merge. I don't see any difference with conda.org's setup.

**Pull request merged**